### DISCUSSION TITLE
Protocol Spec

### DISCUSSION THREAD

**Discussion by tthiery** [smr99/lego-hub-tk#26] [General]
Hi,

  I am the maintainer of [sharpbrick/powered-up](https://github.com/sharpbrick/powered-up) which is a .NET library for communicating with the Lego Technic Powered Up hubs. In that world, I am used to write software on a PC/Mac/Linux device and instruct the hub what to do. I worked hard with the LWP2 spec and collaborated in the community to get the best experience out of it.

  So the question to the Spike/Mindstorm Hub experts: Is the remote protocol of the mindstorms hub a debugging protocol (aka receiving telemetry and messing with the execution of a previous transferred python program) OR is it a true remote control protocol where commands are sent by a controlling PC and then executed by the hub?

I have not found a lot of documentation, so i am bit confused? Also there is a mystical hub to hub protocol which I also found no real information about it.

**Comment by SpudGunMan** [comment]
the Hub is a STM32 board ( think ESP32 ) running micropython as its loader/os/firmware you can use LWP2 functions and then go further with using UART, additionally on the STM32 for example  is a bluetooth /USB ports which can be used. 

with micropython comes REPL and console access that will deliver data that is telemetry sending break commands will interrupt for a interactive shell where you can access the file system etc.. (rshell is a common phrase here) which allows debugging and hardware access

this project aids in the ability for commands to be send PC2Hub

this project and others aim to fix the other issue you have ... there is all DIY right now early stages I hang out on discord:mindstorm hacks and am tracking/working on this project and others

**Reply by tthiery** [reply]
Thanks a lot. I think you confirmed my suspicion ... I have to see what I will do

**Comment by SpudGunMan** [comment]
see the See the [API Design documentation](../blob/main/design.md) for more details on this specific project
